src/schemaVisitor.ts

```diff
@@ -0,0 +1,464 @@
+            // parent .field and grandparent .objectType. Remember that the
+            // current GraphQLSchema is always available via this.schema.
+            field,
+            objectType: type,
+          });
+        });
+      }
+    });
+  }
+
+  visit(schema);
+}
+
+// This class represents a reusable implementation of a @directive that may
+// appear in a GraphQL schema written in Schema Definition Language.
+//
+// By overriding one or more visit{Object,Union,...} methods, a subclass
+// registers interest in certain schema types, such as GraphQLObjectType,
+// GraphQLUnionType, etc. When SchemaDirectiveVisitor.visitSchemaDirectives is
+// called with a GraphQLSchema object and a map of visitor subclasses, the
+// overidden methods of those subclasses allow the visitors to obtain
+// references to any type objects that have @directives attached to them,
+// enabling visitors to inspect or modify the schema as appropriate.
+//
+// For example, if a directive called @rest(url: "...") appears after a field
+// definition, a SchemaDirectiveVisitor subclass could provide meaning to that
+// directive by overriding the visitFieldDefinition method (which receives a
+// GraphQLField parameter), and then the body of that visitor method could
+// manipulate the field's resolver function to fetch data from a REST endpoint
+// described by the url argument passed to the @rest directive:
+//
+//   const typeDefs = `
+//   type Query {
+//     people: [Person] @rest(url: "/api/v1/people")
+//   }`;
+//
+//   const schema = makeExecutableSchema({ typeDefs });
+//
+//   SchemaDirectiveVisitor.visitSchemaDirectives(schema, {
+//     rest: class extends SchemaDirectiveVisitor {
+//       visitFieldDefinition(field: GraphQLField<any, any>) {
+//         const { url } = this.args;
+//         field.resolve = () => fetch(url);
+//       }
+//     }
+//   });
+//
+// The subclass in this example is defined as an anonymous class expression,
+// for brevity. A truly reusable SchemaDirectiveVisitor would most likely be
+// defined in a library using a named class declaration, and then exported for
+// consumption by other modules and packages.
+//
+// See below for a complete list of overridable visitor methods, their
+// parameter types, and more details about the properties exposed by instances
+// of the SchemaDirectiveVisitor class.
+
+export class SchemaDirectiveVisitor extends SchemaVisitor {
+  // The name of the directive this visitor is allowed to visit (that is, the
+  // identifier that appears after the @ character in the schema). Note that
+  // this property is per-instance rather than static because subclasses of
+  // SchemaDirectiveVisitor can be instantiated multiple times to visit
+  // directives of different names. In other words, SchemaDirectiveVisitor
+  // implementations are effectively anonymous, and it's up to the caller of
+  // SchemaDirectiveVisitor.visitSchemaDirectives to assign names to them.
```

Comment:
This is a pretty important comment, too. Essentially, implementors of `SchemaDirectiveVisitor`s don't have to commit to any specific names for the directives they're implementing.